File: backend/orchestrator/zeroclaw_bus.py
"""
ZeroClaw shared state bus — Python backend mirror.
Keeps canonical agent state that endpoints can read/write.
Thread-safe; frontend polls /api/state or subscribes to SSE /api/events.
"""
import threading, time
from dataclasses import dataclass, field, asdict
from typing import Any, Callable


@dataclass
class BusState:
    # Agent 1 outputs
    attention_score:    float = 1.0
    drowsiness_score:   float = 0.0
    drowsiness_flag:    bool  = False
    ear:                float = 0.30
    perclos:            float = 0.0
    blink_freq:         float = 15.0
    # Agent 5 outputs
    fatigue_forecast:   float = 0.0
    forecast_confidence:float = 0.0
    # Enriched context (from Agent 5)
    recommended_rest:   str   = ""
    route_complexity:   str   = "unknown"
    driver_fatigue_state: float = 0.0
    fatigue_forecast_t15: float = 0.0
    # Agent 7 state
    agent7_escalate:    bool  = False
    # NPU health
    npu_status:         str   = "nominal"
    npu_temp_c:         float = 45.0
    npu_ber:            float = 0.0
    npu_latency_dev_ms: float = 0.0
    npu_prediction:     dict  = field(default_factory=dict)
    # Network
    network_state:      str   = "online"
    battery_level:      float = 1.0
    # Session
    session_events:     list  = field(default_factory=list)
    last_sync_ts:       float = 0.0
    next_sync_in_s:     float = 300.0
    # Proactive alert
    proactive_alert_msg:  str = ""
    proactive_alert_urgency: str = ""
    # Driving simulator state (sim ↔ app linkage)
    game_speed:        float = 0.0
    game_rpm:          float = 800.0
    game_location:     str   = "Gachibowli"
    game_segment_type: str   = "parking"
    game_engine_temp:  float = 0.35
    game_fuel:         float = 1.0
    game_oil_pressure: float = 1.0
    game_battery:      float = 1.0
    game_distance_km:  float = 0.0
    game_paused:       bool  = True
    # FL aggregator state (Track 16)
    fl_threshold:      float = 0.70
    fl_last_update:    dict  = field(default_factory=dict)


class ZeroClawBus:
    def __init__(self):
        self._state  = BusState()
        self._lock   = threading.RLock()
        self._subs:  dict[str, list[Callable]] = {}
        self._safety_active = False
        self._sse_clients: list = []     # (queue,) tuples for SSE

    # ------------------------------------------------------------------
    def publish(self, event: str, data: Any):
        _safety_passthrough = (
            "SAFETY_ALERT", "PERCEPTION_UPDATE",
            "GAME_JUMP_REQUEST", "GAME_SPEED_REQUEST",
            "GAME_WEATHER_REQUEST", "GAME_TOGGLE_REQUEST",
        )
        if self._safety_active and event not in _safety_passthrough:
            return
        with self._lock:
            self._apply(event, data)
            serialised = {"event": event, "data": data, "ts": time.time()}
        for handler in self._subs.get(event, []):
            try:
                handler(data)
            except Exception:
                pass
        self._push_sse(serialised)

    def subscribe(self, event: str, handler: Callable):
        self._subs.setdefault(event, []).append(handler)

    def get_state(self) -> dict:
        with self._lock:
            return asdict(self._state)

    def add_sse_client(self, q):
        self._sse_clients.append(q)

    def remove_sse_client(self, q):
        self._sse_clients = [c for c in self._sse_clients if c is not q]

    # ------------------------------------------------------------------
    def _push_sse(self, msg: dict):
        import json
        dead = []
        for q in self._sse_clients:
            try:
                q.put_nowait(json.dumps(msg))
            except Exception:
                dead.append(q)
        for d in dead:
            self.remove_sse_client(d)

    def _apply(self, event: str, data: Any):
        s = self._state
        if event == "PERCEPTION_UPDATE" and isinstance(data, dict):
            s.attention_score  = data.get("attention_score", s.attention_score)
            s.drowsiness_score = data.get("drowsiness_score", s.drowsiness_score)
            s.ear              = data.get("ear", s.ear)
            s.perclos          = data.get("perclos", s.perclos)
            s.blink_freq       = data.get("blink_freq", s.blink_freq)
            if s.drowsiness_score > 0.7 and not self._safety_active:
                self._activate_safety()
                s.drowsiness_flag = True
            elif s.drowsiness_score < 0.5:
                self._safety_active = False
                s.drowsiness_flag = False

        elif event == "FATIGUE_FORECAST" and isinstance(data, dict):
            s.fatigue_forecast    = data.get("fatigue_forecast", 0.0)
            s.forecast_confidence = data.get("forecast_confidence", 0.0)
            ctx = data.get("enriched_system_prompt", {})
            s.driver_fatigue_state  = ctx.get("driver_fatigue_state", s.driver_fatigue_state)
            s.fatigue_forecast_t15  = ctx.get("fatigue_forecast_t15", s.fatigue_forecast_t15)
            s.recommended_rest      = ctx.get("recommended_rest", "") or ""
            s.route_complexity      = ctx.get("route_complexity", s.route_complexity)
            alert = data.get("proactive_alert", {})
            s.proactive_alert_msg     = alert.get("msg", "") or ""
            s.proactive_alert_urgency = alert.get("urgency", "") or ""
            hw = data.get("hardware_health", {})
            s.npu_status = hw.get("status", "nominal")
            s.last_sync_ts   = time.time()
            s.next_sync_in_s = 300.0

        elif event == "SYNC_TICK":
            s.next_sync_in_s = max(0, s.next_sync_in_s - 1)

        elif event == "NPU_HEALTH":
            if isinstance(data, dict):
                s.npu_temp_c       = data.get("temp_c", s.npu_temp_c)
                s.npu_ber          = data.get("ber", s.npu_ber)
                s.npu_latency_dev_ms = data.get("latency_dev_ms", s.npu_latency_dev_ms)

        elif event == "NPU_PREDICTION":
            if isinstance(data, dict):
                s.npu_prediction = data

        elif event == "SESSION_EVENT":
            s.session_events.append(data)
            if len(s.session_events) > 50:
                s.session_events = s.session_events[-50:]

        elif event == "FL_THRESHOLD_UPDATE" and isinstance(data, dict):
            s.fl_threshold = data.get("new_threshold", s.fl_threshold)
            s.fl_last_update = data

        elif event == "GAME_STATE" and isinstance(data, dict):
            s.game_speed         = data.get("game_speed", s.game_speed)
            s.game_rpm           = data.get("game_rpm", s.game_rpm)
            s.game_location      = data.get("game_location", s.game_location)
            s.game_segment_type  = data.get("game_segment_type", s.game_segment_type)
            s.game_engine_temp   = data.get("game_engine_temp", s.game_engine_temp)
            s.game_fuel          = data.get("game_fuel", s.game_fuel)
            s.game_oil_pressure  = data.get("game_oil_pressure", s.game_oil_pressure)
            s.game_battery       = data.get("game_battery", s.game_battery)
            s.game_distance_km   = data.get("game_distance_km", s.game_distance_km)
            s.game_paused        = data.get("game_paused", s.game_paused)

    def _activate_safety(self):
        self._safety_active = True
        self._state.drowsiness_flag = True
        self._push_sse({"event": "SAFETY_ALERT",
                        "data": {"msg": "DROWSINESS DETECTED — ALL AGENTS SUSPENDED"},
                        "ts": time.time()})
        t = threading.Timer(30.0, self._try_deactivate_safety)
        t.daemon = True
        t.start()

    def _try_deactivate_safety(self):
        if self._state.drowsiness_score < 0.5:
            self._safety_active = False
            self._state.drowsiness_flag = False


# Singleton instance shared across all Flask request threads
bus = ZeroClawBus()

File: frontend/js/orchestrator/zeroclaw_bus.js
/**
 * ZeroClaw bus — browser-side shared state and event pub/sub.
 * Mirrors the backend ZeroClawBus. Syncs with server via SSE + REST.
 */
class ZeroClawBus extends EventTarget {
    constructor() {
        super();
        this.state = {
            attention_score:      1.0,
            drowsiness_score:     0.0,
            drowsiness_flag:      false,
            ear:                  0.30,
            perclos:              0.0,
            blink_freq:           15.0,
            fatigue_forecast:     0.0,
            forecast_confidence:  0.0,
            recommended_rest:     '',
            route_complexity:     'unknown',
            driver_fatigue_state: 0.0,
            fatigue_forecast_t15: 0.0,
            proactive_alert_msg:  '',
            proactive_alert_urgency: '',
            npu_status:           'nominal',
            npu_temp_c:           45.0,
            npu_ber:              0.0,
            npu_latency_dev_ms:   0.0,
            network_state:        'online',
            battery_level:        1.0,
            session_events:       [],
            last_sync_ts:         0,
            next_sync_in_s:       300,
            agent7_escalate:      false,
        };
        this._safetyActive = false;
        this._sseSource    = null;
        this._frameQueue   = [];
        this._flushTimer   = null;
    }

    publish(event, data) {
        const _safetyPassthrough = new Set([
            'SAFETY_ALERT', 'PERCEPTION_UPDATE',
            'GAME_JUMP_REQUEST', 'GAME_SPEED_REQUEST',
            'GAME_WEATHER_REQUEST', 'GAME_TOGGLE_REQUEST',
        ]);
        if (this._safetyActive && !_safetyPassthrough.has(event)) return;

        this._apply(event, data);
        this.dispatchEvent(new CustomEvent(event, { detail: data }));
        this.dispatchEvent(new CustomEvent('STATE_CHANGE', { detail: this.state }));

        // Forward perception updates to backend (batched)
        if (event === 'PERCEPTION_UPDATE') {
            this._queueFrameFlush(data);
        }
    }

    subscribe(event, handler) {
        this.addEventListener(event, e => handler(e.detail));
    }

    getState() { return { ...this.state }; }

    // ----------------------------------------------------------------
    // SSE connection to backend
    // ----------------------------------------------------------------
    connectSSE(baseUrl = 'http://localhost:5000') {
        this._baseUrl = baseUrl;
        const src = new EventSource(`${baseUrl}/api/events`);
        src.onmessage = (e) => {
            try {
                const msg = JSON.parse(e.data);
                if (msg.event && msg.data !== undefined) {
                    this._apply(msg.event, msg.data);
                    this.dispatchEvent(new CustomEvent('STATE_CHANGE', { detail: this.state }));
                }
            } catch (_) {}
        };
        src.onerror = () => {
            setTimeout(() => this.connectSSE(baseUrl), 3000);
        };
        this._sseSource = src;

        // Drive the sync countdown locally — backend SYNC_TICK is internal only
        setInterval(() => {
            this._apply('SYNC_TICK', {});
            this.dispatchEvent(new CustomEvent('SYNC_TICK', { detail: {} }));
        }, 1000);
    }

    // ----------------------------------------------------------------
    // Batch-flush perception frames to backend (max 5/s to avoid overload)
    // ----------------------------------------------------------------
    _queueFrameFlush(data) {
        this._frameQueue.push(data);
        if (!this._flushTimer) {
            this._flushTimer = setTimeout(() => {
                const frames = this._frameQueue.splice(0);
                this._flushTimer = null;
                if (frames.length > 0 && this._baseUrl) {
                    fetch(`${this._baseUrl}/api/state/update`, {
                        method: 'POST',
                        headers: { 'Content-Type': 'application/json' },
                        body: JSON.stringify(frames[frames.length - 1]),
                    }).catch(() => {});
                }
            }, 200);   // flush every 200 ms
        }
    }

    // ----------------------------------------------------------------
    _apply(event, data) {
        const s = this.state;
        if (event === 'PERCEPTION_UPDATE' && data) {
            s.attention_score  = data.attention_score  ?? s.attention_score;
            s.drowsiness_score = data.drowsiness_score ?? s.drowsiness_score;
            s.ear              = data.ear              ?? s.ear;
            s.perclos          = data.perclos          ?? s.perclos;
            s.blink_freq       = data.blink_freq       ?? s.blink_freq;

            if (s.drowsiness_score > 0.7 && !this._safetyActive) {
                this._activateSafety();
            } else if (s.drowsiness_score < 0.5) {
                this._safetyActive   = false;
                s.drowsiness_flag    = false;
            }
        }

        if (event === 'FATIGUE_FORECAST' && data) {
            s.fatigue_forecast    = data.fatigue_forecast    ?? 0;
            s.forecast_confidence = data.forecast_confidence ?? 0;
            const ctx = data.enriched_system_prompt || {};
            s.driver_fatigue_state  = ctx.driver_fatigue_state  ?? s.driver_fatigue_state;
            s.fatigue_forecast_t15  = ctx.fatigue_forecast_t15  ?? s.fatigue_forecast_t15;
            s.recommended_rest      = ctx.recommended_rest       ?? '';
            s.route_complexity      = ctx.route_complexity       ?? s.route_complexity;
            const alert = data.proactive_alert || {};
            s.proactive_alert_msg     = alert.msg     ?? '';
            s.proactive_alert_urgency = alert.urgency ?? '';
            const hw = data.hardware_health || {};
            s.npu_status = hw.status ?? 'nominal';
            s.last_sync_ts   = Date.now() / 1000;
            s.next_sync_in_s = 300;
        }

        if (event === 'SYNC_TICK') {
            s.next_sync_in_s = Math.max(0, s.next_sync_in_s - 1);
        }

        if (event === 'SESSION_EVENT' && data) {
            s.session_events = [...s.session_events.slice(-49), data];
        }

        if (event === 'SAFETY_ALERT') {
            s.drowsiness_flag = true;
        }
    }

    _activateSafety() {
        this._safetyActive    = true;
        this.state.drowsiness_flag = true;
        this.dispatchEvent(new CustomEvent('SAFETY_ALERT',
            { detail: { msg: 'DROWSINESS DETECTED', urgency: 'critical' } }));
        setTimeout(() => {
            if (this.state.drowsiness_score < 0.5) {
                this._safetyActive = false;
                this.state.drowsiness_flag = false;
            }
        }, 30000);
    }
}

window.ZeroClawBus = ZeroClawBus;
